lib/screens/v2/candidate_generator/helpers/alerts.ex

```diff
@@ -0,0 +1,179 @@
+            Access.filter(&(&1["type"] == "trip")),
+            "relationships",
+            "stops",
+            "data",
+            Access.all(),
+            "id"
+          ])
+
+        {:ok, stop_sequences}
+
+      _ ->
+        :error
+    end
+  end
+
+  @spec fetch_routes_at_stop(String.t(), Date.t()) ::
+          {:ok, list(%{route_id: String.t(), active?: boolean()})} | :error
+  def fetch_routes_at_stop(
+        stop_id,
+        today,
+        fetch_all_route_ids_fn \\ &fetch_all_route_ids/1,
+        fetch_active_route_ids_fn \\ &fetch_active_route_ids/2
+      ) do
+    with {:ok, all_route_ids} <- fetch_all_route_ids_fn.(stop_id),
+         {:ok, active_route_ids} <- fetch_active_route_ids_fn.(stop_id, today) do
+      active_set = MapSet.new(active_route_ids)
+
+      routes_at_stop =
+        Enum.map(all_route_ids, &%{route_id: &1, active?: MapSet.member?(active_set, &1)})
+
+      {:ok, routes_at_stop}
+    else
+      :error -> :error
+    end
+  end
+
+  @spec fetch_alerts(list(String.t()), list(String.t())) :: {:ok, list(Alert.t())} | :error
+  def fetch_alerts(
+        stop_ids,
+        route_ids,
+        fetch_fn \\ &Alert.fetch/1,
+        filter_fn \\ &filter_alerts/3
+      ) do
+    with {:ok, stop_based_alerts} <- fetch_fn.(stop_ids: stop_ids, route_ids: route_ids),
+         {:ok, route_based_alerts} <- fetch_fn.(route_ids: route_ids) do
+      merged_alerts =
+        [stop_based_alerts, route_based_alerts]
+        |> Enum.concat()
+        |> Enum.uniq_by(& &1.id)
+        |> filter_fn.(stop_ids, route_ids)
+
+      {:ok, merged_alerts}
+    else
+      :error -> :error
+    end
+  end
+
+  @doc """
+  Filters out alerts whose effects we are not interested in, as well as those that do not inform at least one of:
+  - an entire route type, e.g. bus or light rail
+  - a route that serves the home stop, scoped to either the home stop or a downstream stop
+  - a downstream stop or the home stop
+  - a route that serves the home stop
+
+  (list describes the `relevant_ie?` function clauses in order)
+  """
+  def filter_alerts(alerts, stop_ids, route_ids) do
+    stop_id_set = MapSet.new(stop_ids)
+    route_id_set = MapSet.new(route_ids)
+
+    relevant_ie? = fn
+      %{route_type: route_type, stop: nil, route: nil} when not is_nil(route_type) ->
+        true
+
+      %{stop: stop, route: route} when not is_nil(stop) and not is_nil(route) ->
+        stop in stop_id_set and route in route_id_set
+
+      %{stop: stop, route: nil} ->
+        stop in stop_id_set
+
+      %{stop: nil, route: route} ->
+        route in route_id_set
+
+      _ ->
+        false
+    end
+
+    effects = relevant_effects()
+
+    alerts
+    |> Enum.filter(&(&1.effect in effects))
+    |> Enum.filter(&Enum.any?(&1.informed_entities, relevant_ie?))
+  end
+
+  defp relevant_effects do
+    MapSet.new(
+      ~w[shuttle station_closure stop_closure suspension detour stop_moved snow_route elevator_closure]a
+    )
```

Comment:
We've had issues in the past with Dialyzer producing mysterious complaints about MapSets stored as module attributes—[something about opaque types](https://elixirforum.com/t/can-mapset-be-an-attribute/34681)—but I'll try again in case that issue's been fixed since we last tried.

Reply:
The Dialyzer issue is still there, but it doesn't complain as long as I don't explicitly call a `MapSet` function on the set.

`&(&1.effect in @relevant_effects)` -> no issue
`&MapSet.member?(@relevant_effects, &1.effect)` -> opaque type complaint 🙃

Reply:
Can you add the warning to `.dialyzer-ignore-warnings`?  [API example](https://github.com/mbta/api/blob/master/.dialyzer.ignore-warnings#L12)

Reply:
Yeah, I think that's what I'd do if there's an easy way to target this specific warning regardless of where it might appear.